File: app/src/main/java/org/dhis2/usescases/datasets/datasetDetail/DataSetDetailPresenter.java
package org.dhis2.usescases.datasets.datasetDetail;

import static org.dhis2.commons.matomo.Actions.GRANULAR_SYNC;
import static org.dhis2.commons.matomo.Actions.OPEN_ANALYTICS;
import static org.dhis2.commons.matomo.Categories.DATASET_LIST;
import static org.dhis2.commons.matomo.Labels.CLICK;

import androidx.annotation.VisibleForTesting;

import org.dhis2.commons.schedulers.SchedulerProvider;
import org.dhis2.commons.filters.data.FilterRepository;
import org.dhis2.commons.filters.DisableHomeFiltersFromSettingsApp;
import org.dhis2.commons.filters.FilterItem;
import org.dhis2.commons.filters.FilterManager;
import org.dhis2.commons.matomo.MatomoAnalyticsController;
import org.hisp.dhis.android.core.organisationunit.OrganisationUnit;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.disposables.CompositeDisposable;
import timber.log.Timber;

public class DataSetDetailPresenter {

    private DataSetDetailView view;
    private DataSetDetailRepository dataSetDetailRepository;
    private SchedulerProvider schedulerProvider;
    private FilterManager filterManager;
    private FilterRepository filterRepository;
    private DisableHomeFiltersFromSettingsApp disableHomFilters;
    private MatomoAnalyticsController matomoAnalyticsController;

    CompositeDisposable disposable;

    public DataSetDetailPresenter(DataSetDetailView view,
                                  DataSetDetailRepository dataSetDetailRepository,
                                  SchedulerProvider schedulerProvider,
                                  FilterManager filterManager,
                                  FilterRepository filterRepository,
                                  DisableHomeFiltersFromSettingsApp disableHomFilters,
                                  MatomoAnalyticsController matomoAnalyticsController
    ) {

        this.view = view;
        this.dataSetDetailRepository = dataSetDetailRepository;
        this.schedulerProvider = schedulerProvider;
        this.filterManager = filterManager;
        this.filterRepository = filterRepository;
        this.disableHomFilters = disableHomFilters;
        this.matomoAnalyticsController = matomoAnalyticsController;
        disposable = new CompositeDisposable();
    }

    public void init() {
        getOrgUnits();

        disposable.add(
                filterManager.asFlowable().startWith(filterManager)
                        .flatMap(filterManager -> Flowable.just(filterRepository.dataSetFilters(dataSetDetailRepository.getDataSetUid())))
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribe(filterItems -> {
                                    if (filterItems.isEmpty()) {
                                        view.hideFilters();
                                    } else {
                                        view.setFilters(filterItems);
                                    }
                                    view.updateFilters(filterManager.getTotalFilters());
                                },
                                Timber::d
                        )
        );

        disposable.add(
                filterManager.getPeriodRequest()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribe(
                                periodRequest -> view.showPeriodRequest(periodRequest.getFirst()),
                                Timber::e
                        ));


        disposable.add(FilterManager.getInstance().getCatComboRequest()
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                        catComboUid -> view.showCatOptComboDialog(catComboUid),
                        Timber::e
                )
        );
    }

    public void onSyncClicked(){
        view.showGranularSync();
    }

    public void onBackClick() {
        view.back();
    }

    public void showFilter() {
        view.showHideFilter();
    }

    @VisibleForTesting
    public void getOrgUnits() {
        disposable.add(
                filterManager.ouTreeFlowable()
                        .subscribeOn(schedulerProvider.io())
                        .observeOn(schedulerProvider.ui())
                        .subscribe(
                                open -> view.openOrgUnitTreeSelector(),
                                Timber::e
                        )
        );
    }


    public void onDettach() {
        disposable.clear();
    }

    public void displayMessage(String message) {
        view.displayMessage(message);
    }

    public void clearFilterClick() {
        filterManager.clearAllFilters();
        view.clearFilters();
    }

    public void filterCatOptCombo(String selectedCatOptionCombo) {
        FilterManager.getInstance().addCatOptCombo(
                dataSetDetailRepository.getCatOptCombo(selectedCatOptionCombo)
        );
    }

    public void clearFilterIfDatasetConfig() {
        List<FilterItem> filters = filterRepository.homeFilters();
        disableHomFilters.execute(filters);
    }

    public void setOpeningFilterToNone() {
        filterRepository.collapseAllFilters();
    }

    public void setOrgUnitFilters(List<OrganisationUnit> selectedOrgUnits) {
        FilterManager.getInstance().addOrgUnits(selectedOrgUnits);
    }

    public void refreshList() {
        filterManager.publishData();
    }

    public void trackDataSetAnalytics() {
        matomoAnalyticsController.trackEvent(DATASET_LIST, OPEN_ANALYTICS, CLICK);
    }

    public void trackDataSetGranularSync() {
        matomoAnalyticsController.trackEvent(DATASET_LIST,GRANULAR_SYNC, CLICK );
    }
}

File: app/src/main/java/org/dhis2/usescases/datasets/datasetDetail/DataSetDetailPresenter.kt
/*
* Copyright (c) 2004-2019, University of Oslo
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
* Redistributions of source code must retain the above copyright notice, this
* list of conditions and the following disclaimer.
*
* Redistributions in binary form must reproduce the above copyright notice,
* this list of conditions and the following disclaimer in the documentation
* and/or other materials provided with the distribution.
* Neither the name of the HISP project nor the names of its contributors may
* be used to endorse or promote products derived from this software without
* specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
* ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
* ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

package org.dhis2.usescases.datasets.datasetDetail

import androidx.annotation.VisibleForTesting
import io.reactivex.disposables.CompositeDisposable
import org.dhis2.data.schedulers.SchedulerProvider
import org.dhis2.utils.filters.FilterManager
import timber.log.Timber

class DataSetDetailPresenter internal constructor(
    private val view: DataSetDetailView,
    private val dataSetDetailRepository: DataSetDetailRepository,
    private val schedulerProvider: SchedulerProvider,
    private val filterManager: FilterManager
) {
    var disposable: CompositeDisposable = CompositeDisposable()

    fun init() {
        getOrgUnits()

        disposable.add(
            filterManager.asFlowable()
                .startWith(FilterManager.getInstance())
                .flatMap { filterManager ->
                    dataSetDetailRepository.dataSetGroups(
                        filterManager.orgUnitUidsFilters,
                        filterManager.periodFilters,
                        filterManager.stateFilters,
                        filterManager.catOptComboFilters
                    )
                }
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { data ->
                        view.setData(data)
                        view.updateFilters(filterManager.totalFilters)
                    },
                    { Timber.d(it) }
                )
        )

        disposable.add(
            filterManager.periodRequest
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { periodRequest -> view.showPeriodRequest(periodRequest) },
                    { Timber.e(it) }
                )
        )

        disposable.add(
            dataSetDetailRepository.catOptionCombos()
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { view.setCatOptionComboFilter(it) },
                    { Timber.e(it) }
                )
        )

        disposable.add(
            dataSetDetailRepository.canWriteAny()
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { view.setWritePermission(it) },
                    { Timber.e(it) }
                )
        )
    }

    fun addDataSet() {
        view.startNewDataSet()
    }

    fun onBackClick() {
        view.back()
    }

    fun openDataSet(dataSet: DataSetDetailModel) {
        view.openDataSet(dataSet)
    }

    fun showFilter() {
        view.showHideFilter()
    }

    @VisibleForTesting
    fun getOrgUnits() {
        disposable.add(
            filterManager.ouTreeFlowable()
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(
                    { view.openOrgUnitTreeSelector() },
                    { Timber.e(it) }
                )
        )
    }

    fun onDettach() {
        disposable.clear()
    }

    fun displayMessage(message: String) {
        view.displayMessage(message)
    }

    fun onSyncIconClick(dataSet: DataSetDetailModel) {
        view.showSyncDialog(dataSet)
    }

    fun updateFilters() {
        filterManager.publishData()
    }

    fun clearFilterClick() {
        filterManager.clearAllFilters()
        view.clearFilters()
    }
}
